execution/executor/src/lib.rs

```diff
@@ -381,19 +383,48 @@ where
         )?;
 
         info!("Skipping the first {} transactions.", num_txns_to_skip);
+        let txn_list_is_empty = txn_list_with_proof.is_empty();
         let transactions: Vec<_> = txn_list_with_proof
             .transactions
             .into_iter()
             .skip(num_txns_to_skip as usize)
             .collect();
 
         // If the proof is verified, then the length of txn_infos and txns must be the same.
-        let transaction_infos: Vec<_> = txn_list_with_proof
-            .proof
-            .transaction_infos()
-            .iter()
-            .skip(num_txns_to_skip as usize)
-            .collect();
+        let (skipped_transaction_infos, transaction_infos) = {
+            let (a, b) = txn_list_with_proof
+                .proof
+                .transaction_infos()
+                .split_at(num_txns_to_skip as usize);
+            (a.to_vec(), b.to_vec())
+        };
+
+        // verify no fork happens.
+        if !txn_list_is_empty {
+            let frozen_subtree_roots_from_proof = txn_list_with_proof
+                .proof
+                .left_siblings()
+                .iter()
+                .rev()
+                .cloned()
+                .collect::<Vec<_>>();
+            let accu_from_proof = InMemoryAccumulator::<TransactionAccumulatorHasher>::new(
+                frozen_subtree_roots_from_proof,
+                first_version - num_txns_to_skip,
+            )?
+            .append(
+                &skipped_transaction_infos
+                    .iter()
+                    .map(CryptoHash::hash)
+                    .collect::<Vec<_>>()[..],
+            );
+
+            // The two accumulator root hashes should be identical.
+            ensure!(
+                self.cache.synced_trees().state_id() == accu_from_proof.root_hash(),
+                "Fork happens because the current synced_trees doesn't match the txn list provided."
+            )
+        }
```

Comment:
extract to separate method for readability?

Reply:
I am a little confused about what the inputs should be if get separated. maybe needs more refactoring.